WIP lifetime categories

mgehre opened the pull request:
My goal is to implement parsing, diagnostics and documentation for gsl::Owner and gsl::Pointer attributes in that
branch and then push that to LLVM phabricator before the next step.

mgehre committed:
WIP lifetime categories

mgehre commented:
Solved '  File /home/gehre/llvm-project/clang/test/SemaCXX/attr-gsl-owner-pointer.cpp Line 24: expected '(' for function-style cast or type construction'

Xazax-hun commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -331,6 +331,13 @@ unsigned Parser::ParseAttributeArgsCommon(
+  if (attributeIsTypeArgAttr(*AttrName)) {
+    ParseAttributeWithTypeArg(*AttrName, AttrNameLoc, Attrs, EndLoc, ScopeName,
+                              ScopeLoc, Syntax);
+    return 1;
```
Is this the number of arguments? Can we be sure all of the type argument attributes have only one argument? Maybe we should at least leave a FIXME.

Xazax-hun commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -7110,6 +7143,12 @@ static void ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D,
+  case ParsedAttr::AT_Owner:
+    handleLifetimeCategoryAttr(S, D, AL);
+    break;
+  case ParsedAttr::AT_Pointer:
```
Maybe we could have a fallthrough case here to avoid repetition?

Xazax-hun commented on clang/test/AST/ast-dump-attr.cpp:
```diff
@@ -211,6 +211,15 @@ namespace TestSuppress {
+namespace TestLifetimeCategories {
+  class [[gsl::Owner(int)]] AOwner {};
+  // CHECK: CXXRecordDecl{{.*}} struct AOwner
+  // CHECK-NEXT: ObjCBridgeRelatedAttr{{.*}} NSParagraphStyle
```
What is this `ObjCBridgeRelatedAttr`?

Xazax-hun commented on clang/test/SemaCXX/attr-gsl-owner-pointer.cpp:
```diff
@@ -0,0 +1,36 @@
+class [[gsl::Pointer("int")]] PointerDerefNoType {};
+
+
+class [[gsl::Owner(int)]] [[gsl::Pointer(int)]] BothOwnerPointer {};
```
I think we should diagnose this as an error. A TODO is also fine if we plan to do it in a later commit.

Xazax-hun commented on clang/test/SemaCXX/attr-gsl-owner-pointer.cpp:
```diff
@@ -0,0 +1,36 @@
+
+
+class [[gsl::Owner(7)]] OwnerDerefNoType {};
+class [[gsl::Pointer("int")]] PointerDerefNoType {};
```
Maybe we should add a TODO if this is not diagnosed yet.

Xazax-hun reviewed (commented):
Thanks, great work, it is looking good. Some minor nits inline.

mgehre committed:
More test fixes

mgehre committed:
Fix more tests and review comments

mgehre commented on clang/test/SemaCXX/attr-gsl-owner-pointer.cpp:
```diff
@@ -0,0 +1,55 @@
+class [[gsl::Pointer(S)]] APointer {};
+
+class AddOwnerLater {};
+class [[gsl::Owner(int)]] AddOwnerLater;
```
@Xazax-hun: This feature allows us to ship a "annotation header" like
```
namespace std {
template<class T>
class [[gsl::Owner(T)]] vector;
}
```
to add annotations to existing code.

Xazax-hun commented on clang/test/SemaCXX/attr-gsl-owner-pointer.cpp:
```diff
@@ -0,0 +1,55 @@
+class AddOwnerLater {};
+class [[gsl::Owner(int)]] AddOwnerLater;
+
+// TODO: diagnose me
```
Maybe we can diagnose this in `handleLifetimeCategoryAttr`? Trying to get the conflicting attribute from the `Decl` right before allocating the parsed one?

Xazax-hun committed:
Disallow conflicting lifetime attributes on later decls.

Xazax-hun committed:
Add preliminary documentation to attributes.

Xazax-hun committed:
Fix some review comments.

mgehre committed:
Add lifetime categories attributes

Summary:
This is the first part of work announced in
"[RFC] Adding lifetime analysis to clang" [0],
i.e. the addition of the [[gsl::Owner(T)]] and
[[gsl::Pointer(T)]] attributes, which
will enable user-defined types to participate in
the lifetime analysis (which will be part of the
next PR).
The type `T` here is called "DerefType" in the paper,
and denotes the type that an Owner owns and a Pointer
points to. E.g. `std::vector<int>` should be annotated
with `[[gsl::Owner(int)]]` and
a `std::vector<int>::iterator` with `[[gsl::Pointer(int)]]`.

We explicitly allow to add an annotation after
the definition of the class to allow adding annotations
to external source from by the user, e.g.
```

namespace std {
template<typename T, typename Alloc>
class [[gsl::Owner(T)]] vector;
}
```
until we can ship a standard library with annotations.

[0] http://lists.llvm.org/pipermail/cfe-dev/2018-November/060355.html

Reviewers: gribozavr

Subscribers: xazax.hun, cfe-commits

Tags: #clang

Differential Revision: https://reviews.llvm.org/D63954

Xazax-hun committed:
Address some review comments.

Xazax-hun committed:
Address more review comments.

mgehre committed:
git-clang-format

mgehre committed:
Add newline at end of file

mgehre committed:
gsl::Owner/gsl::Pointer: Make DerefType parameter optional

mgehre committed:
Don't crash AST dump when Owner/Pointer has no Type argument

Xazax-hun commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -2766,6 +2766,20 @@ def TypeTagForDatatype : InheritableAttr {
+}
+
+def Pointer : InheritableAttr {
+  let Subjects = SubjectList<[CXXRecord]>;
```
Usually spellings comes first and subjects later.

Xazax-hun committed:
Make the list of attribute prperties more consistent.

mgehre committed:
Don't crash when pretty-printing an Attribute with optional type argument

mgehre committed:
Unify type and non-type attribute parsing; don't add OwnerAttr/PointerAttr twice

mgehre committed:
Improve documentation

mgehre committed:
Disallow reference and array types as DerefType

mgehre committed:
Add example to documentation.

mgehre committed:
Diagnose unions; improve other diagnostics; fix all comments

mgehre committed:
Fix all comments

mgehre committed:
Fix 80 col

Xazax-hun commented:
No longer relevant as this work has already been upstreamed. Nice job!
